### PR TITLE
Optimisations on compression in package build / lrzip package

### PR DESCRIPTION
**Pull request by Wetitpig**
As title, the number of threads can be passed through `XZ_OPT` to achieve multithreading compression during packaging for debian.

### CODE REVIEW

**Comment by Wetitpig** [comment]
By the way, although gzipping in Debian packages seems to fail for ARM processors, should we allow the user to choose the compression format just like for pacman packages?

**Comment by Wetitpig** [comment]
`zstd` compresses much faster and with a slightly higher ratio than `gzip`.

File: scripts/setup-termux.sh [hunk 1]
```diff
@@ -5,10 +5,14 @@ PACKAGES=""
 PACKAGES+=" clang"		# Required for termux-elf-cleaner and C/C++ packages.
 PACKAGES+=" file"		# Used in termux_step_massage().
 PACKAGES+=" gnupg"		# Used in termux_get_repo_files() and build-package.sh.
+PACKAGES+=" lrzip"
```
Comment (xtkoba) [inline scripts/setup-termux.sh:8]: Please add a comment about this package to align with others. The same goes for `lzop` and `lz4` below.

File: packages/lrzip/configure.ac.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/configure.ac b/configure.ac
+index 4b12022..4714ef4 100644
+--- a/configure.ac
++++ b/configure.ac
+@@ -108,7 +108,7 @@ AC_C_INLINE
+ AC_C_BIGENDIAN
+ 
+ AC_CHECK_LIB(pthread, pthread_create, ,
+-        AC_MSG_ERROR([Could not find pthread library - please install libpthread]))
++        AC_MSG_NOTICE([Could not find pthread library - please install libpthread]))
```
Comment (xtkoba) [inline packages/lrzip/configure.ac.patch:10]: Bionic libc does not have `libpthread.*` files. You can omit the entire `AC_CHECK_LIB` code here.

**Review by xtkoba — COMMENTED** [review]
Please squash the commits appropriately. (Squashing into one does not seem to be approproate.)

**Comment by xtkoba** [comment]
As pointed out in https://github.com/termux/termux-packages/pull/9124#issuecomment-1050017038, this may increase the size of binary packages. Some profiling is needed. Rebuild some typical packages from small to large with this configuration and show that the size increase is negligible.

**Comment by xtkoba** [comment]
Also, it will be nice if use of multi-threading can be opt-out, or rather, opt-in.

**Comment by Wetitpig** [comment]
Just added the option for enabling multithreading. i.e. it is disabled by default, which seems safer

File: .github/workflows/packages.yml [hunk 3]
```diff
@@ -188,7 +188,7 @@ jobs:
         done
 
         if [ ! -z "$packages" ]; then
-          ./scripts/run-docker.sh ./build-package.sh -I -a ${{ matrix.target_arch }} $packages
+          ./scripts/run-docker.sh ./build-package.sh -I -a ${{ matrix.target_arch }} --multixz $packages
```
Comment (xtkoba) [inline .github/workflows/packages.yml:191]: This should not be done by default, if you are aiming for safety.

File: scripts/build/termux_create_debian_subpackages.sh [hunk 4]
```diff
@@ -59,7 +59,11 @@ termux_create_debian_subpackages() {
 		cd "$SUB_PKG_DIR/massage"
 		local SUB_PKG_INSTALLSIZE
 		SUB_PKG_INSTALLSIZE=$(du -sk . | cut -f 1)
-		tar -cJf "$SUB_PKG_PACKAGE_DIR/data.tar.xz" .
+		if [ -z "${TERMUX_PACKAGE_FORMAT-}" ]; then
```
Comment (xtkoba) [inline scripts/build/termux_create_debian_subpackages.sh:62]: Presumably should be `TERMUX_XZ_MULTITHREAD`.

File: .github/workflows/packages.yml (hunk 3 above)
Comment (Wetitpig) [reply .github/workflows/packages.yml:191]: But then most builds are carried out on Github Actions and I think they impose the time limit...

**Comment by xtkoba** [comment]
So this adds a global variable `TERMUX_XZ_MULTITHREAD`. AFAIK, our convention is to always define a global variable. You should add

```bash
TERMUX_XZ_MULTITHREAD=false
```

somewhere in `scripts/build/termux_step_setup_variables.sh`.

And then the check should be something like

```bash
if [ "$TERMUX_XZ_MULTITHREAD" = "true" ]; then
	: # some operation
fi
```

File: .github/workflows/packages.yml (hunk 3 above)
Comment (xtkoba) [reply .github/workflows/packages.yml:191]: Is the time limit significant? I would say it is negligible in most cases. At most only a handful of packages will really need the multi-threading.

Comment (xtkoba) [reply .github/workflows/packages.yml:191]: And for such a package it should be sufficient to specify the flag in its `build.sh`. This change in the workflow is not necessary at all.

**Comment by Wetitpig** [comment]
Instead of checking at every invocation of `xz` and `zstd` I would rather set the corresponding environment variables. 

I am not sure if the command for `lrzip` in `scripts/build/termux_step_create_pacman_package.sh` should be changed too.

**Comment by xtkoba** [comment]
CC: @Maxython (as pacman maintainer)

File: .github/workflows/packages.yml [hunk 5]
```diff
@@ -188,7 +188,7 @@ jobs:
         done
 
         if [ ! -z "$packages" ]; then
-          ./scripts/run-docker.sh ./build-package.sh -I -a ${{ matrix.target_arch }} $packages
+          ./scripts/run-docker.sh ./build-package.sh -I -a ${{ matrix.target_arch }} --multiz $packages
```
Comment (xtkoba) [inline .github/workflows/packages.yml:191]: This affects every package. Such a change should not be made before enough tests. Please be more conservative.

I don't find any profiling report yet. Is this useful for us in the first place?

Comment (Wetitpig) [reply .github/workflows/packages.yml:191]: Now checking on `libllvm`

Other packages with sources distributed as `xz` may also be tested

Comment (Wetitpig) [reply .github/workflows/packages.yml:191]: I am encountering difficulties with rebuilding the docker image or `zstd` and then building a package in the same pipe.
Is there a way to do so on Github Actions? Thanks

File: scripts/setup-termux.sh (hunk 1 above)
Comment (Wetitpig) [reply scripts/setup-termux.sh]: Resolved in latest commit

File: .github/workflows/packages.yml (hunk 5 above)
Comment (Grimler91) [reply .github/workflows/packages.yml:191]: Only way would probably be to fork the docker image and install the extra packages in it (easier to test locally, but I understand that might not be an option).

I pushed an update to add those additional packages to the docker image, so should be possible to test once build for https://github.com/termux/termux-packages/commit/6e61626b6c319a26dd92e46b844645c70f585042 finishes

File: .github/workflows/packages.yml (hunk 3 above)
Comment (Wetitpig) [reply .github/workflows/packages.yml]: Benchmark with `libllvm`:

Size in MB:
| Arch | Time with `--multixz` | Size with `--multixz` | without `--multixz` | Size without `--multixz` |
| --- | --- | --- | --- | --- |
| aarch64 | 2:39 | 144 | 2:39 | 143 |
| arm | 2:40 | 148 | 3:38 | 146 |
| i686 | 2:41 | 162 | 3:47 | 160 |
| x86_64 | 2:46 | 18 | 2:40 | 156 |

**Comment by Wetitpig** [comment]
How about keeping the option while removing the argument from CI?

**Pull request closed without merging**